pkg/controlplane/client.go

```diff
@@ -78,16 +90,90 @@ func (c *client) Authorize(req *api.AuthorizationRequest) (*remoteServerAuthoriz
 	return resp, nil
 }
 
+// IsActive returns if the peer is active or not.
+func (c *client) IsActive() bool {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	return c.active
+}
+
+// setActive the peer status (active or not).
+func (c *client) setActive(active bool) {
+	c.lock.Lock()
+	defer c.lock.Lock()
+	c.active = active
+	if active || c.lastSeen.IsZero() {
+		c.lastSeen = time.Now()
+	}
+}
+
+// GetHeartbeat get a heartbeat from other peers.
+func (c *client) getHeartbeat() error {
+	var retErr error
+	// copy peer clients array aside
+	peerClients := make([]*jsonapi.Client, len(c.clients))
+	{
+		c.lock.RLock()
+		defer c.lock.RUnlock()
+		copy(peerClients, c.clients)
+	}
+
+	for _, client := range peerClients {
+		serverResp, err := client.Get(api.HeartbeatPath)
+		if err != nil {
+			retErr = errors.Join(retErr, err)
+		} else {
+			if serverResp.Status == http.StatusOK {
+				return nil
+			}
+			retErr = errors.Join(retErr, fmt.Errorf("unable to get heartbeat (%d), server returned: %s",
+				serverResp.Status, serverResp.Body))
+		}
+	}
+
+	return retErr // Return an error if all client targets are unreachable
+}
+
+// heartbeatMonitor checks all peers for responsiveness, every fixed amount of time.
+func (c *client) heartbeatMonitor() {
+	c.logger.Info("Starting sending heartbeats to peer")
+	ticker := time.NewTicker(heartbeatInterval)
+	defer ticker.Stop()
+	// get heartbeat from all the peers
+	for {
+		t := time.Now()
+		if c.IsActive() || (!c.IsActive() && (t.Sub(c.lastSeen) > heartbeatRetransmissionTime)) ||
+			c.lastSeen.IsZero() {
+			if err := c.getHeartbeat(); err != nil {
+				if c.IsActive() {
+					c.logger.Errorf("Unable to get heartbeat from peer  Error: %v", err.Error())
+					c.setActive(false)
+				}
+			} else {
+				c.setActive(true)
+			}
+		}
+
+		// wait till it's time for next heartbeat round
+		<-ticker.C
+	}
+}
+
 // newClient returns a new Peer API client.
 func newClient(peer *store.Peer, tlsConfig *tls.Config) *client {
 	clients := make([]*jsonapi.Client, len(peer.Gateways))
 	for i, endpoint := range peer.Gateways {
 		clients[i] = jsonapi.NewClient(endpoint.Host, endpoint.Port, tlsConfig)
 	}
-	return &client{
-		clients: clients,
+	c := &client{
+		clients:  clients,
+		active:   false,
+		lastSeen: time.Time{},
 		logger: logrus.WithFields(logrus.Fields{
 			"component": "peer-client",
 			"peer":      peer}),
 	}
+
+	go c.heartbeatMonitor()
```

Comment:
this go routine does not terminate, correct? What if the peer is removed?

Reply:
So you remove the client and the go routine is part of the client so it will be terminated.